subsys/bluetooth/host/gatt.c

```diff
@@ -5210,6 +5227,71 @@ int bt_gatt_store_ccc(uint8_t id, const bt_addr_le_t *addr)
 	return 0;
 }
 
+/** Struct used to store both the id and the random address of a device when replacing
+ * random addresses in the ccc attribute's cfg array with the device's id address after
+ * pairing complete.
+ */
+struct addr_match {
+	const bt_addr_le_t *private_addr;
+	const bt_addr_le_t *id_addr;
+};
+
+uint8_t convert_to_id_on_match(const struct bt_gatt_attr *attr,
+							   uint16_t handle, void *user_data)
+{
+	struct _bt_gatt_ccc *ccc;
+	struct addr_match *device_addresses = user_data;
+
+	/* Check if attribute is a CCC */
+	if (attr->write != bt_gatt_attr_write_ccc) {
+		return BT_GATT_ITER_CONTINUE;
+	}
+
+	ccc = attr->user_data;
+
+	/* Copy the device's id address to the config's address if the config's address is the
+	 * same as the device's private address
+	 */
+	for (size_t i = 0; i < ARRAY_SIZE(ccc->cfg); i++) {
+		if (bt_addr_le_cmp((const bt_addr_le_t *)&ccc->cfg[i].peer,
+		    device_addresses->private_addr) == 0) {
+			bt_addr_le_copy(&ccc->cfg[i].peer, device_addresses->id_addr);
+		}
+	}
+
+	return BT_GATT_ITER_CONTINUE;
+}
+
+void bt_gatt_update_ccc_cfg_addr(const bt_addr_le_t *private_addr, const bt_addr_le_t *id_addr)
+{
+	struct addr_match user_data = {
+		.private_addr = private_addr,
+		.id_addr	  = id_addr
+	};
+
+	bt_gatt_foreach_attr(0x0001, 0xffff, convert_to_id_on_match, &user_data);
+}
+
+int bt_gatt_store_ccc_and_cf(struct bt_conn *conn)
+{
+	int err = bt_gatt_store_ccc(conn->id, &(conn->le.dst));
+
+	if (err != 0) {
+		return err;
+	}
+
+	err = bt_gatt_store_cf(conn);
+
+	return err;
+}
+
+void bt_gatt_identity_resolved(struct bt_conn *conn, const bt_addr_le_t *private_addr,
+			       const bt_addr_le_t *id_addr)
+{
+	bt_gatt_update_ccc_cfg_addr(private_addr, id_addr);
+	bt_gatt_store_ccc_and_cf(conn);
```

Comment:
Since these are both single-purpose functions called from a single place and both are fairly short, I'd just code them inline directly into this function. Btw, remember to declare functions as `static` if they're only used within a single c-file. Also place new static functions as high up in the c-file as possible in order to avoid the need for forward-declarations (i.e. if possible, place this before the point where you register the callbacks).